components/locale/src/serde/langid.rs

```diff
@@ -0,0 +1,63 @@
+use crate::LanguageIdentifier;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+impl Serialize for LanguageIdentifier {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for LanguageIdentifier {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct LanguageIdentifierVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for LanguageIdentifierVisitor {
+            type Value = LanguageIdentifier;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(formatter, "A valid Unicode Language Identifier")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                s.parse::<LanguageIdentifier>()
+                    .map_err(serde::de::Error::custom)
```

Comment:
I'm not sure. @XAMPPRocky - do you have any thoughts on this? I don't think we had to implement anything concrete here.

Reply:
The custom here is from serde::de::Error not icu-locale which accepts virtually any error type. https://docs.rs/serde/1.0.110/serde/de/trait.Error.html#tymethod.custom

You could also use `unexpected_value` but it will only change the error message, it’s pretty common to just pass the error into custom and leave it at that.